packages/amplify_core/lib/src/types/query/query_field_operators.dart

```diff
@@ -180,11 +180,19 @@ class ContainsQueryOperator extends QueryFieldOperatorSingleValue<String> {
       : super(value, QueryFieldOperatorType.contains);
 
   @override
-  bool evaluate(String? other) {
+  bool evaluate(dynamic other) {
     if (other == null) {
```

Comment:
We are excluding `contains(null)` is because `[String]` and `[String!]` in model schema are both generated as `List<String>` in flutter correct?

Reply:
The null check here isn't excluding `other.contains(null)` from being called. It is preventing `null.contains(value)` from being called. `other` is the field on the model being compared, and `value` is the value passed into the predicate.

However, `other.contains(null)` will not be called. This is because `value` is a non-nullable `String`. That makes sense for string comparisons. `'foo'.contains(null)` would throw an exception. However, `['foo', null].contains(null)` is valid.

I can test what happens if value is type `String?`, and a null value is provided. I a not sure if it will work if codegen is typing `[String]` as `List<String>` as you said. I am also not sure if this currently works for .query(), although it seems like it should.

On a similar topic, I originally changed `value` from `String` to `Comparable` to support calling contains() for any `List<T>` where `T` is Comparable. This would allow the contains operator to work for not only `[String]`, but also `[Int]`, `[Float]`, `[AWSDateTime]`, etc. I left it typed as `String` because .query() does not seem to support this. On the native side (not sure if it is in Amplify-Flutter or Amplify-iOS) the value is getting cast to a String and so it throws an exception.

That seems like a bug to me, but was outside of scope of this PR. For this PR I just want to get `.observeQuery()` on par with `.query()`. Would you agree that is a bug? It seems like .contains() should work for any `List<T>` where `T` is Comparable.

Reply:
To make things more complicated ...

`ModelWithAppsyncScalarTypes.LISTOFINTVALUE.contains('1')` will actually work for .query() even though LISTOFINTVALUE is `List<int>` and `'1'` is a `String`. This will not work for .observeQuery() (with or without this PR). 

`ModelWithAppsyncScalarTypes.INTVALUE.contains('1')` will also work for .query(), but not for .observeQuery().

I don't know if it is intentional that these both work for .query(). It might just be a result of the SQL implementation (probably using a LIKE operator).

We could make both of these work for .observeQuery() as well by doing the following:

```dart
bool evaluate(dynamic other) {
  if (other == null) {
    return false;
  } else if (other is List<String>) {
    return other.contains(value);
  } else if (other is List) {
     // support non-string Lists 
    return other.map((dynamic e) => e.toString()).contains(value);
  } else if (other is String) {
    return other.contains(value);
  } else {
    // support non String values
    return other.toString().contains(value);
  }
}
```

This seems hacky and error prone. I would prefer to leave the implementation in the PR, even though it means there will still be a difference in support between .query() and .observeQuery() for the .contains() predicate. We can always revisit this if it turns out this functionality is intentional.

I am going to add some of this context in the issue I open.

Reply:
Thanks for the digging and testing!

>I manually changed the generated models to allow this, and it caused the app to crash.

I think this is expected as we've started some work enabling `[String?]` some time last year but never finished, there is a draft PR for this...

>This would allow the contains operator to work for not only [String], but also [Int], [Float], [AWSDateTime]

I'm not sure if it's necessary to enable `contains` on other types other than `String`, as AppSync `contains` filter supports only `String` and `ID` (which is `String` in client) types. 

My comments are not blocking merging this PR.